src/TraceEvent/Computers/ContentionTimeComputer.cs

```diff
@@ -0,0 +1,63 @@
+﻿using System.Collections.Generic;
+using Microsoft.Diagnostics.Tracing.Etlx;
+using Microsoft.Diagnostics.Tracing.Parsers.Clr;
+using Microsoft.Diagnostics.Tracing.Stacks;
+
+namespace Microsoft.Diagnostics.Tracing.Computers
+{
+    public class ContentionTimeComputer
+    {
+        private readonly TraceLog _eventLog;
+        private readonly MutableTraceEventStackSource _stackSource;
+        private readonly Dictionary<int, StackSourceSample> _samplesPerThread;
+
+        private const double NanosInMillisecond = 1000 * 1000;
+
+        public ContentionTimeComputer(TraceLog eventLog, MutableTraceEventStackSource stackSource)
+        {
+            _eventLog = eventLog;
+            _stackSource = stackSource;
+            _samplesPerThread = new Dictionary<int, StackSourceSample>();
+        }
+
+        public void GenerateContentionTimeStacks()
+        {
+            var eventSource = _eventLog.Events.GetSource();
+            
+            eventSource.Clr.ContentionStart += OnContentionStart;
+            eventSource.Clr.ContentionStop += OnContentionStop;
+
+            eventSource.Process();
+            
+            _stackSource.DoneAddingSamples();
+        }
+
+        private void OnContentionStart(ContentionStartTraceData startData)
+        {
+            if (!_samplesPerThread.TryGetValue(startData.ThreadID, out var sample))
+            {
+                _samplesPerThread[startData.ThreadID] = sample = new StackSourceSample(_stackSource);
+                sample.Count = 1;
+            }
+
+            sample.TimeRelativeMSec = startData.TimeStampRelativeMSec;
+            var callStackIdx = startData.CallStackIndex();
+            var stackIndex = _stackSource.GetCallStack(callStackIdx, startData);
+            sample.StackIndex = stackIndex;
+        }
+
+        private void OnContentionStop(ContentionStopTraceData stopData)
+        {
+            if (!_samplesPerThread.TryGetValue(stopData.ThreadID, out var sample))
+            {
+                // no corresponding start event
+                return;
+            }
+
+            sample.Metric = (float) (stopData.DurationNs / NanosInMillisecond);
+            sample.StackIndex = _stackSource.Interner.CallStackIntern(_stackSource.Interner.FrameIntern($"EventData DurationNs {stopData.DurationNs:N0}"), sample.StackIndex);
+
+            _stackSource.AddSample(sample);
```

Comment:
After adding the sample, please clear the `StackIndex`, `Metric`, and `TimeRelativeMSec`.  Then, you should also look for these unset values in `OnContentionStop` and bail out early.  This will handle the case where you see a previous `Contention/Start` that paired properly with a `Contention/Stop`, but then where a `Contention/Start` was missed.  In that case, the current code would add the time range from the first `Contention/Start` to the first `Contention/Stop` and then the time from the first `Contention/Start` to the second `Contention/Stop`.

Reply:
I did that but there are other cases that I'm not sure how to treat meaningfully. For example, we could receive Start -> Start sequence and the current code will just silently drop the first Start. I wonder if I should signal it somehow to the user to indicate the data is flawed? However in this case it'd mean that something is very wrong either with clr or with parsers, not sure what user can do about that.

Reply:
Yes, that is a risk.  It can show up in a couple of ways:

1. The trace ends before the End event gets fired.
2. The trace is missing events.

Generally speaking, I don't worry about this, and take one of two approaches:

1. Ignore the incomplete time ranges.
2. Annotate the incomplete time ranges in a way that lets them be filtered out separately if need be via ExcPats.

You can also partially solve the problem by looking for starts that never have another start or stop event during the trace, and setting the stop time to be the end of the trace.  This accounts for starts that are missing stops due to completion of the trace (case 1 above).

For this PR, I'm OK either way.  Missing events aren't super common, and we can often capture new traces when this happens.  It's the cases where the trace starts or ends in the middle of an operation that's much more common.

Reply:
I rewrote the code to handle all of these cases and also to remove the duplication between Contention and WaitHandleWait. 

Special cases are handled as following:
* Stop is the first event of a thread: not an error, set metric to `timestamp of stop - start of the trace`
* Start is the last event of a thread: not an error, set metric to `end of trace - timestamp of start`
* Start -> Start sequence: error, add `BROKEN_EVENT NO_CORRESPONDING_STOP` to the stack, set metric to `float.Epsilon`
* Stop -> Stop sequence: error, add `BROKEN_EVENT NO_CORRESPONDING_START` to the stack, set metric to `float.Epsilon`